shared/js/background/components/toggle-reports.js

```diff
@@ -0,0 +1,265 @@
+
+        registerMessageHandler(
+            'getToggleReportOptions',
+            (_, sender) => this.toggleReportStarted(sender)
+        )
+
+        registerMessageHandler(
+            'rejectToggleReport',
+            (_, sender) => this.toggleReportFinished(false, sender)
+        )
+
+        registerMessageHandler(
+            'sendToggleReport',
+            (_, sender) => this.toggleReportFinished(true, sender)
+        )
+
+        createAlarm(ToggleReports.ALARM_NAME, { periodInMinutes: 60 })
+        browser.alarms.onAlarm.addListener(async ({ name }) => {
+            if (name === ToggleReports.ALARM_NAME) {
+                await ToggleReports.clearExpiredResponses()
+            }
+        })
+    }
+
+    /**
+     * Provides details of what will be included in a breakage report, so that
+     * the user can make an informed decision. Called when the "toggle reports"
+     * UI flow begins.
+     *
+     * @param {browser.Runtime.Port} sender
+     * @returns {Promise<ToggleReportOptions>}
+     */
+    async toggleReportStarted (sender) {
+        // If the browser closes the popup UI during the "toggle reports" flow
+        // (e.g. because the user clicked away), the connection will drop and
+        // this event will fire.
+        sender?.onDisconnect?.addListener(this.onDisconnect)
+
+        let siteUrl = null
+        const currentTabUrl = (await getCurrentTab())?.url
+        if (currentTabUrl) {
+            siteUrl = getURLWithoutQueryString(currentTabUrl)
+        }
+
+        /** @type {ToggleReportOptions} */
+        const response = { data: [] }
+
+        for (const paramId of ToggleReports.PARAM_IDS) {
+            if (paramId === 'siteUrl' && siteUrl) {
+                response.data.push(
+                    { id: 'siteUrl', additional: { url: siteUrl } }
+                )
+            } else {
+                response.data.push({ id: paramId })
+            }
+        }
+
+        return response
+    }
+
+    /**
+     * Called when the "toggle reports" UI flow finishes. If the user chose to
+     * send the breakage report that will be sent now.
+     *
+     * @param {boolean} accepted
+     *   True if the user opted to send the breakage report, false otherwise.
+     * @param {browser.Runtime.Port?} sender
+     * @returns {Promise<void>}
+     */
+    async toggleReportFinished (accepted, sender) {
+        sender?.onDisconnect?.removeListener(this.onDisconnect)
+
+        // Note the response and time.
+        await settings.ready()
+        const times = settings.getSetting('toggleReportTimes') || []
+        times.push({ timestamp: Date.now(), accepted })
+        settings.updateSetting('toggleReportTimes', times)
+
+        if (accepted) {
+            await sendBreakageReportForCurrentTab({
+                pixelName: 'protection-toggled-off-breakage-report',
+                reportFlow: 'on_protections_off_dashboard_main'
+            })
+
+            await resolveAfterDelay(2000)
```

Comment:
Is this delay part of the spec for this flow, or a number we've picked to make the flow feel nicer?

Reply:
Without the delay, the "Thanks for the report" message won't be visible. It's not spec'd as far as I know, but we are discussing how long the delay should be exactly in the ship review.